### PR TITLE
feat: Implement Mutating Webhooks

### PR DESCRIPTION
**Pull request by chzar**
#### Overview

The most basic types of automation Kubernetes admins are interested in are validating and mutating webhooks. Currently shell-operator only supports validating webhooks. Because validating and mutating webhooks are nearly identical except for the patch field, the existing validating webhook manager can be expanded to also support mutating webhooks.

#### What this PR does / why we need it

Closes https://github.com/flant/shell-operator/issues/325

#### Special notes for your reviewer

Status:
- support for mutating webhooks has been added to the webhook manager.
- The validating package and types have been renamed to 'admission' to reflect their more general usage.
- Remainder of the codebase has been updated with the new package names and types.

TODO: 
- hook controller fixup
- 'activate' mutating webhook configs and add test coverage for mutating webhooks. 
- end-to-end testing in my own k8s environment
- documentation

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

**Comment by chzar** [comment]
still a lot of work to be done to properly route events. Just tracing the code paths and making changes accordingly.

**Comment by chzar** [comment]
Ok it should be working at this point - I can trigger the webhook by calling the http api manually - then shell-operator crashes when trying do a GET request to the k8s api after running the test script.

I think this is expected behaviour at this point because I dont think this functionality can be tested outside of a cluster. Would like to hear from other contributors about this.

File: pkg/webhook/admission/types/response.go [hunk 1]
```diff
@@ -10,13 +10,14 @@ import (
 	"strings"
 )
 
-type ValidatingResponse struct {
+type AdmissionResponse struct {
 	Allowed  bool     `json:"allowed"`
 	Message  string   `json:"message,omitempty"`
 	Warnings []string `json:"warnings,omitempty"`
+	Patch    []byte   `json:"patch,omitempty" protobuf:"bytes,4,opt,name=patch"`
```
Comment (diafour) [inline pkg/webhook/admission/types/response.go:17]:
```suggestion
	Patch    []byte   `json:"patch,omitempty"`
```

**Comment by diafour** [comment]
Hello! Great work!

I've copied example/204-validating-webhook and created a hook with `kubernetesMutating` configuration in #443. And it works!

The hook is simple, it mutates a field in the CronTab custom resource:
```
  PATCH=$( echo '[{"op": "add", "path": "/spec/replicas", "value": 333 }]' | base64 )
    cat <<EOF > $VALIDATING_RESPONSE_PATH
{"allowed":true, "patch": "$PATCH"}
EOF
```

There is an important addition to make it actualy work: "copy Patch from the hook response to the AdmissionResponse": https://github.com/flant/shell-operator/commit/b5e7272020f91454fdb58038eb9ea64e353975a2. You can cherry-pick it to your branch.

There are some minor issues:

1. An `event` label in logs is still "kubernetesValidating":

```
{"level":"info","msg":"Got AdmissionReview request for confId='hooks' webhookId='change-crontab-resources'","time":"2022-12-08T06:34:01Z"}
{"binding":"kubernetesValidating","event.id":"63149ed4-c66c-439e-bd75-e9e613248603","level":"debug","msg":"Handle 'kubernetesValidating' event 'hooks' 'change-crontab-resources'","time":"2022-12-08T06:34:01Z"}
{"binding":"change.crontab.resources","event":"kubernetesValidating","hook":"mutating.sh","level":"info","msg":"Execute hook","queue":"","task":"HookRun","time":"2022-12-08T06:34:01Z"}
{"binding":"change.crontab.resources","event":"kubernetesValidating","hook":"mutating.sh","level":"debug","msg":"snapshot info: no kubernetes bindings for hook","queue":"","task":"HookRun","time":"2022-12-08T06:34:01Z"}
{"binding":"change.crontab.resources","event":"kubernetesValidating","hook":"mutating.sh","level":"debug","msg":"Executing command '/hooks/mutating.sh' in '/hooks' dir","queue":"","task":"HookRun","time":"2022-12-08T06:34:01Z"}
```

2. No `__on_mutating` handler in the shell framework https://github.com/flant/shell-operator/blob/main/frameworks/shell/hook.sh#L54-L59 

3. There is an extra field in MutatingWebhookConfiguration: `reinvocationPolicy`. https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#mutatingwebhook-v1-admissionregistration-k8s-io (a link to all fields, there is no direct link)
It ok to add it in a separate PR.

**Comment by chzar** [comment]
@diafour
 
Thanks Ivan!

I have been a bit short on time lately and was struggling to get minikube and hyperv networking to play nice so I could do some testing. I will try and address your comments push some code as soon as I can!

**Comment by chzar** [comment]
I have set this PR as ready to review. There is still some additional needs for tests/docs, but I can take that in a follow up.

**Comment by diafour** [comment]
Thank you @chzar ! I've run '206 mutating example' with your branch and everything works as expected. It's time to merge this great work.

**Pull request merged**